### PR TITLE
Adddsyev

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

**Comment by btracey** [comment]
Needs https://github.com/gonum/lapack/pull/107

File: lapack.go [hunk 1]
```diff
@@ -109,3 +111,12 @@ const (
 	SVDOverwrite        = 'O' // Compute the singular vectors and store them in input matrix
 	SVDNone             = 'N' // Do not compute singular vectors
 )
+
+// EigComp specifies the type of eigenvalue decomposition.
+type EigComp byte
+
+const (
+	EigValueOnly EigComp = 'N' // Only compute eigenvalues.
+	EigDecomp            = 'V' // Finds eigenvectors of decomposed matrix.
```
Comment (kortschak) [inline lapack.go]: s/Finds/Find/

File: lapack.go [hunk 2]
```diff
@@ -109,3 +111,12 @@ const (
...
+	EigBoth              = 'I' // Eigenvalues of the tridiagonal matrix.
```
Comment (kortschak) [inline lapack.go]: Find eigenvalues of the...

File: native/dsyev.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+		down
+		up
+	)
+	var sigma float64
+	iscale := none
+	if anrm > 0 && anrm < rmin {
+		iscale = down
+		sigma = rmin / anrm
+	} else if anrm > rmax {
+		iscale = up
```
Comment (kortschak) [inline native/dsyev.go]: The fortran marks this is down (1) - identical to the case above.

File: native/dsyev.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	if !wantz {
+		ok = impl.Dsterf(n, w, work[indE:])
+	} else {
+		impl.Dorgtr(uplo, n, a, lda, work[indTau:], work[indWrk:], lwork2)
+		ok = impl.Dsteqr(jobz, n, w, work[indE:], a, lda, work[indTau:])
+	}
+
+	bi := blas64.Implementation()
+	// If matrix was scaled, then rescale eigenvalues appropriately.
+	if iscale == down {
```
Comment (kortschak) [inline native/dsyev.go]: up is never considered?

File: native/dsyev.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	// For eigenvalues only, call Dsterf. For eigenvectors, first call Dorgtr to
+	// generate the orthogonal matrix, then call Dsteqr.
+	if !wantz {
+		ok = impl.Dsterf(n, w, work[indE:])
+	} else {
+		impl.Dorgtr(uplo, n, a, lda, work[indTau:], work[indWrk:], lwork2)
+		ok = impl.Dsteqr(jobz, n, w, work[indE:], a, lda, work[indTau:])
+	}
+
+	bi := blas64.Implementation()
```
Comment (kortschak) [inline native/dsyev.go]: This is only used once, so move it into the condition block.

**Comment by kortschak** [comment]
Comments for dsteqr in #107.

File: lapack.go (hunk 1 above)
Comment (btracey) [inline lapack.go]: Fixed in #107

File: lapack.go (hunk 2 above)
Comment (btracey) [inline lapack.go]: Fixed in #107

File: native/dsyev.go (hunk 3 above)
Comment (btracey) [inline native/dsyev.go]: iscale can only be 1 or 0 in this code. Changed to a boolean with a better name.

File: native/dsyev.go (hunk 5 above)
Comment (btracey) [inline native/dsyev.go]: Done.

File: native/dsyev.go (hunk 4 above)
Comment (btracey) [inline native/dsyev.go]: See above.

**Comment by btracey** [comment]
I added Syev to the lapack64 interface and added cgo test.

File: cgo/lapack.go [hunk 6]
```diff
@@ -995,6 +996,40 @@ func (impl Implementation) Dpocon(uplo blas.Uplo, n int, a []float64, lda int, a
...
 }
 
+// Dsyev computes all eigenvalues and, optionally, the eigenvectors of a real
+// symmetric matrix A.
+//
+// w contains the eigenvalues in ascending order upon return. w must have length
+// at least n, and Dsyev will panic otherwise.
+//
+// On entry, a contains the elements of the symmetric matrix A in the triangular
+// portion specified by uplo. If jobz == lapack.EigComp A contains the
```
Comment (vladimir-ch) [inline cgo/lapack.go]: s/EigComp A/EigComp, a/

File: lapack64/lapack64.go [hunk 7]
```diff
@@ -324,6 +324,26 @@ func Pocon(a blas64.Symmetric, anorm float64, work []float64, iwork []int) float
...
 }
 
+// Syev computes all eigenvalues and, optionally, the eigenvectors of a real
+// symmetric matrix A.
+//
+// w contains the eigenvalues in ascending order upon return. w must have length
+// at least n, and Dsyev will panic otherwise.
+//
+// On entry, a contains the elements of the symmetric matrix A in the triangular
+// portion specified by uplo. If jobz == lapack.EigComp A contains the
```
Comment (vladimir-ch) [inline lapack64/lapack64.go]: Also here.

File: cgo/lapack.go (hunk 6 above)
Comment (btracey) [inline cgo/lapack.go]: Done.

File: lapack64/lapack64.go (hunk 7 above)
Comment (btracey) [inline lapack64/lapack64.go]: Done.

File: native/dsyev.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+	}
+	if n == 0 {
+		return true
+	}
+	if n == 1 {
+		w[0] = a[0]
+		work[0] = 2
+		if wantz {
+			a[0] = 1
+			return true
```
Comment (vladimir-ch) [inline native/dsyev.go]: Shouldn't this return be outside of the if scope?

File: testlapack/dsyev.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+	Dsyev(jobz lapack.EigComp, uplo blas.Uplo, n int, a []float64, lda int, w, work []float64, lwork int) (ok bool)
+}
+
+func DsyevTest(t *testing.T, impl Dsyever) {
+	rnd := rand.New(rand.NewSource(1))
+	for _, uplo := range []blas.Uplo{blas.Lower, blas.Upper} {
+		for _, test := range []struct {
+			n, lda int
+		}{
+			{2, 0},
```
Comment (vladimir-ch) [inline testlapack/dsyev.go]: Add tests with size 1?

**Comment by vladimir-ch** [comment]
LGTM

File: native/dsyev.go (hunk 8 above)
Comment (btracey) [inline native/dsyev.go]: Done.

**Pull request merged**